src/Microsoft.OData.Core/UrlValidation/ODataUrlValidationRules/ODataUrlValidationRule.cs

```diff
@@ -0,0 +1,97 @@
+// </copyright>
+//--------
+
+using System;
+
+namespace Microsoft.OData.UriParser.Validation
+{
+    /// <summary>
+    /// Base class for OData URL Validation Rules
+    /// </summary>
+    public abstract class ODataUrlValidationRule
+    {
+        /// <summary>
+        /// True if rule should include implicitly selected properties (i.e., in the absence of a $select)
+        /// </summary>
+        public bool IncludeImpliedPropeties { get; set; }
+
+        /// <summary>
+        /// Name of the rule.
+        /// </summary>
+        public string RuleName { get; protected set; }
+
+        /// <summary>
+        /// Internal API called by the validation engine to validate the item of the specified type
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="component"></param>
+        internal abstract void Validate(ODataUrlValidationContext context, object component);
+
+        /// <summary>
+        /// Internal API called to get the type of Url element that the rule targets
+        /// </summary>
+        /// <returns>The type of the Url element that the rule targets</returns>
+        internal abstract Type GetRuleType();
+    }
+
+    /// <summary>
+    /// Rule for validating all elements of an ODataUrl of a particular type.
+    /// </summary>
+    /// <typeparam name="T">The type of OData Url element that the rule validates.</typeparam>
+    public sealed class ODataUrlValidationRule<T> : ODataUrlValidationRule where T : class
+    {
+        /// <summary>
+        /// The action that validates an instance of the specified type.
+        /// </summary>
+        private Action<ODataUrlValidationContext, T> validateMethod;
+
+        /// <summary>
+        /// The type of the Url element that the rule targets
+        /// </summary>
+        internal Type ruleType = typeof(T);
+
+        /// <summary>
+        /// Constructs an instance of an ODataValdiationRule of a particular type.
+        /// </summary>
+        /// <param name="validateMethod">The Action that validates an instance of the specified type.></param>
+        public ODataUrlValidationRule(string ruleName, Action<ODataUrlValidationContext, T> validateMethod)
+        {
+            this.RuleName = ruleName;
+            this.validateMethod = validateMethod;
+        }
+
+        /// <summary>
+        /// Constructs an instance of an ODataValdiationRule of a particular type.
+        /// </summary>
+        /// <param name="validateMethod">The Action that validates an instance of the specified type.></param>
+        /// <param name="includeImpliedProperties">Whether the rule should include implicitly selected properties.</param>
+        public ODataUrlValidationRule(string ruleName, Action<ODataUrlValidationContext, T> validateMethod, bool includeImpliedProperties)
+        {
+            this.RuleName = ruleName;
+            this.IncludeImpliedPropeties = true;

```

Comment:
includeImpliedProperties is not used in the constructor method?

Reply:
Oops -- good catch, thanks! I think I changed it for testing purposes and never changed it back...

---
In reply to: [434873599](https://github.com/OData/odata.net/pull/1769#discussion_r434873599) [](ancestors = 434873599)